retrybp/retry.go

```diff
@@ -48,10 +48,19 @@ func GetOptions(ctx context.Context) (options []retry.Option, ok bool) {
 //
 // 2. If retry.Do returns a batch of errors (retry.Error), put those in a
 // errorsbp.Batch from baseplate.go.
+//
+// It also auto applies retry.Context with the ctx given,
+// so that the retries will be stopped as soon as ctx is canceled.
+// You can override this behavior by injecting a retry.Context option into ctx.
 func Do(ctx context.Context, fn func() error, defaults ...retry.Option) error {
 	options, _ := GetOptions(ctx)
-	options = append(defaults, options...)
-	err := retry.Do(fn, options...)
+	mergedOptions := make([]retry.Option, 1, 1+len(defaults)+len(options))
+	// Always do this as the first option, to allow overriding it in either
+	// defaults or options from the ctx.
+	mergedOptions[0] = retry.Context(ctx)
+	mergedOptions = append(mergedOptions, defaults...)
+	mergedOptions = append(mergedOptions, options...)
+	err := retry.Do(fn, mergedOptions...)
```

Comment:
❗  This feels like a footgun.

There are a few options, including at least:
* Make it a type-safe positional param.
* Make it able to find it regardless of parameter, usually by making options an interface.
* Apply the options and then update the resulting option struct, and then pass that as a single option.

Reply:
this is limited by the upstream api: https://pkg.go.dev/github.com/avast/retry-go#Do. also this is a common pattern used by go libraries (I don't particularly love this pattern, fwiw).

Reply:
Sorry, I missed that it was just adding the first option, I misread it as _changing_ the first option, assuming it was always something.

🔕